lib/evmmax/evmmax.cpp

```diff
@@ -0,0 +1,114 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2023 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <evmmax/evmmax.hpp>
+
+using namespace intx;
+
+namespace evmmax
+{
+namespace
+{
+inline constexpr uint64_t mul_inv64(uint64_t base) noexcept
+{
+    uint64_t result = 1;
+    for (auto i = 0; i < 64; ++i)
+    {
+        result *= base;
+        base *= base;
+    }
+    return result;
+}
+
+template <typename UintT>
+inline UintT compute_r_squared(const UintT& mod) noexcept
+{
+    static constexpr auto r2 = intx::uint<UintT::num_bits * 2 + 64>{1} << UintT::num_bits * 2;
```

Comment:
hard to remember precedence of `<<` vs `*`
```suggestion
    static constexpr auto r2 = intx::uint<UintT::num_bits * 2 + 64>{1} << (UintT::num_bits * 2);
```